### ISSUE TITLE
Error during template compile of 'MyModule'

### ISSUE THREAD

**Issue by matheushf** [ng-packagr/ng-packagr#607]
## Type of Issue

Error when importing the compiled library

```
[x] Bug Report
[ ] Feature Request
```

## Description

After running ng-packagr, I get the bundle and everything just fine, but when I'm going to import it in my Angular 5 App I get the error:

`ERROR in Error during template compile of 'MyModule'
  Function calls are not supported in decorators in 'NgModule'
    'NgModule' calls a function at @angular/core/core.ts(194,50).
`

I'm importing it like:

`import { CountDownModule, CountDownComponent } from '../../../../projectLab/ng2-date-countdown'`
```
@NgModule({
  declarations: [
    AppComponent,    
    CountDownComponent
  ],
  imports: [
    BrowserModule,
    CountDownModule,
  ],
```

#### How To Reproduce

Build an library with ng-packagr and import it in an Angular app.

#### Expected Behaviour

No errors

#### Version Information

```
ng-packagr: 2.0.0
@angular/*: x.y.z
typescript: x.y.z
rxjs: x.y.z
node: x.y.z
npm/yarn: x.y.z
```

_please include any version information that might be relevant, e.g. other third-party libraries_

**Comment by choucry13** [comment]
Hi, i have the same when i use a static function with a module.

**Comment by deebloo** [comment]
@matheushf can you share your CountDownModule?

**Comment by matheushf** [comment]
```
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CountDownComponent } from './components';

@NgModule({
  declarations: [
    CountDownComponent,
  ],
  exports: [
    CountDownComponent,
  ],
  imports: [
    CommonModule
  ]
})

export class CountDownModule {
}
```

**Comment by about-code** [comment]
@matheushf I can't see any obvious Syntax error and assume your imports are correct. What happens if you remove the trailing commas ? I've found the aot compiler to be pretty nasty about them in the past.

**Comment by dherges** [comment]
Please provide a repro with instructions or a source code repository, if action should be taken.

**Comment by Gbuomprisco** [comment]
Hi everyone, I suspect my package is failing for the same reason. 

The error stacktrace is simply:

```
ERROR in Error during template compile of 'NgxContentLoadingModule'
  Function calls are not supported in decorators in 'NgModule'
    'NgModule' calls a function at ../@angular/core/core.ts(194,31).
```

Not sure why, but my package returns no issues if I link it locally. :thinking:

If anyone wants to take a look at the code, here is the package: https://github.com/Gbuomprisco/ngx-content-loading.

Thanks!

**Comment by about-code** [comment]
Jason Aden suggests [here](https://github.com/angular/angular/issues/20858) for a similar issue to add the `RouterModule` to the `exports` of the feature module. Does this help any further?

**Comment by matheushf** [comment]
@dherges You can get my source code from [here](https://github.com/matheushf/ng2-date-countdown/tree/AOT)

Build it and for testing you can `npm link` it to another project, and run using `--prod` for AOT, then it will chrash.

**Comment by Korigoth** [comment]
any news on this problem?

we are getting this annoying error too

**Comment by stephengardner** [comment]
Bump, sub, having similar issues

**Comment by dherges** [comment]
@Gary-Pan You're shipping `dependencies` and `devDependencies` with the `@my/lib` package. See https://github.com/Gary-Pan/my-lib/blob/master/package.json#L10-L44

@matheushf You're shipping `devDependencies` with your library. The `dist/package.json` is:

```json
{"name":"ng2-date-countdown","version":"0.0.5","description":"Angular2 Simple Date CountDown","main":"bundles/ng2-date-countdown.umd.js","module":"ng2-date-countdown.es5.js","typings":"ng2-date-countdown.d.ts","$schema":"./node_modules/ng-packagr/package.schema.json",
  "dependencies":{},
  "devDependencies":{
     "@angular/common":"5.1.2","@angular/compiler":"^5.1.2","@angular/compiler-cli":"^5.2.4","@angular/core":"5.1.2","@angular/platform-server":"^5.2.4","@ngtools/webpack":"^1.9.4","@types/core-js":"^0.9.43","@types/webpack":"^3.0.13","@types/webpack-env":"^1.13.0","angular2-template-loader":"^0.6.2","compression-webpack-plugin":"^1.1.6","copy-webpack-plugin":"^4.4.1","css-loader":"^0.28.9","extract-text-webpack-plugin":"^3.0.2","file-loader":"^1.1.6","gulp":"^3.9.1","gulp-inline-ng2-template":"^4.1.0","html-loader":"^0.5.5","html-webpack-plugin":"^2.30.1","json-loader":"^0.5.7","ng-packagr":"^1.7.0","node-sass":"^4.7.2","raw-loader":"^0.5.1","rxjs":"^5.5.6","sass-loader":"^6.0.6","to-string-loader":"^1.1.5","typescript":"^2.7.1","webpack-angular-externals":"^1.0.2","webpack-merge":"^4.1.1","webpack-rxjs-externals":"^1.1.0","zone.js":"^0.8.20"},"es2015":"ng2-date-countdown.js","metadata":"ng2-date-countdown.metadata.json"}
```

**Comment by dherges** [comment]
What do your resolved `node_modules` in the application look like?

What can easily happen with `dependencies` and `devDependencies` is that you end up with duplicated modules, e.g.:

```bash
node_modules
  |- @angular
      |- common
  |- @my
      |- lib
          |- node_modules
              |- @angular
                  |- common
```

You should be using `peerDependencies` to say what your library depends on.

---

For ng-packagr:
 - should it auto-generate the `peerDependencies` section at build time?
 - should it build error by default when `dependencies` (other than tslib) and `devDependencies` are in package.json and make it optional to suppress this error?

**Comment by Gary-Pan** [comment]
@dherges I resolved it by replace the root package.json with dist/package.json. I forgot that I installed @my/lib with git repo. So, it should only retain the dist folder. Thanks for you help.

**Comment by matheushf** [comment]
Thanks for your answer @dherges 

I thought it would work, but it sadly didn't. I'm testing it from local, importing it like '../../../project-folder'. 
I changed from devDependencies to peerDependencies and again still I get the errors:

```
ERROR in : Unexpected value 'CountDownComponent in /var/www/labs/projectLab/ng2-date-countdown/dist/src/components/countdown/countdown.component.d.ts' 
declared by the module 'AppModule in /var/www/labs/testLab/app-ng2-date-countdown/src/app/app.module.ts'. 
Please add a @Pipe/@Directive/@Component annotation.
```

```
ERROR in Error during template compile of 'CountDownModule'
  Function expressions are not supported in decorators in 'NgModule'
    'NgModule' contains the error at ../@angular/core/core.ts(194,31)
      Consider changing the function expression into an exported function.
```

**Comment by noelmace** [comment]
I also have the same issue, when migrating an existing (an working) library from a custom build system to ng-packagr.

Output : 
```
ERROR in Error during template compile of 'AppModule'
  Function calls are not supported in decorators but 'MyLibCookieModule' was called
```

My package.json file in my node_modules/@mylib/common looks like this, so I don't think the issue is here : 
```
{
  "$schema": "../../node_modules/ng-packagr/package.schema.json",
  "name": "@mylib/common",
  "version": "3.0.0-alpha.1",
  "peerDependencies": {
    "@angular/common": "^5.2.1",
    "@angular/core": "^5.2.1",
    "@angular/forms": "^5.2.1",
    "@angular/http": "^5.2.1",
    "@angular/platform-browser": "^5.2.1",
    "@angular/platform-browser-dynamic": "^5.2.1",
    "@angular/router": "^5.2.1",
    "@ng-bootstrap/ng-bootstrap": "^1.0.0",
    "@ngrx/effects": "^4.1.1",
    "@ngrx/router-store": "^4.1.1",
    "@ngrx/store": "^4.1.1",
    "bootstrap": "^4.0.0",
    "core-js": "^2.5.3",
    "rxjs": "^5.5.6",
    "stacktrace-js": "^2.0.0",
    "zone.js": "^0.8.20"
  },
  "publishConfig": {
    "registry": "http://myorg.net/repository/myrepo/"
  },
  "main": "bundles/mylib-common.umd.js",
  "module": "esm5/mylib-common.js",
  "es2015": "esm2015/mylib-common.js",
  "typings": "mylib-common.d.ts",
  "metadata": "mylib-common.metadata.json",
  "dependencies": {
    "tslib": "^1.7.1"
  }
}

```

The consumer project's AppModule : 
```
import {
  MyLibCookieModule,
} from '@atlas/common';

// ...

const imports = [
  BrowserModule,
  FormsModule,
  HttpClientModule,
  MyLibCookieModule.forRoot(),
];

@NgModule({
  declarations: [AppComponent],
  providers: [
      // ...
  ],
  imports: imports,
  bootstrap: [AppComponent]
})
export class AppModule {}
``` 

tsconfig.app.json in the consumer project : 
```
{
  "compilerOptions": {
    "alwaysStrict": true,
    "baseUrl": "",
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "noStrictGenericChecks": true,
    "lib": [
      "es2017",
      "dom"
    ],
    "module": "es2015",
    "moduleResolution": "node",
    "outDir": "../out-tsc/app",
    "sourceMap": true,
    "target": "es5",
    "types": [
      "node",
      "google-maps",
      "googlemaps"
    ],
    "typeRoots": [
      "../node_modules/@types"
    ]
  },
  "exclude": [
    "test.ts",
    "testing",
    "**/*.spec.ts",
    "**/*.stub.ts",
    "**/*.mock.ts"
  ]
}
```

In the lib, the cookie.module.ts file is : 
```
import { ModuleWithProviders, NgModule } from '@angular/core';
import { CookieOptionsProvider, COOKIE_OPTIONS } from './cookie-options-provider';
import { CookieOptions } from './cookie-options.model';
import { cookieServiceFactory } from './cookie.factory';
import { CookieService } from './cookie.service';

@NgModule({ providers: [CookieOptionsProvider] })
export class MyLibCookieModule {
  /**
   * Use this method in your root module to provide the CookieService
   * @param options
   * @returns
   */
  static forRoot(options: CookieOptions = {}): ModuleWithProviders {
    return {
      ngModule: MyLibCookieModule,
      providers: [
        { provide: COOKIE_OPTIONS, useValue: options },
        { provide: CookieService, useFactory: cookieServiceFactory, deps: [CookieOptionsProvider] }
      ]
    };
  }

  /**
   * Use this method in your other (non root) modules to import the directive/pipe
   * @param options
   * @returns
   */
  static forChild(options: CookieOptions = {}): ModuleWithProviders {
    return {
      ngModule: MyLibCookieModule,
      providers: [
        { provide: COOKIE_OPTIONS, useValue: options },
        { provide: CookieService, useFactory: cookieServiceFactory, deps: [CookieOptionsProvider] }
      ]
    };
  }
}
```

**Comment by SirWojtek** [comment]
I had the same issue:
* the library package was successfully compiled and bundled
* everything was working on `jit` build
* an `aot` compilation gives me the message: `Function calls are not supported in decorators`

In my case, the root cause was missing decorators metadata in the library bundle (`*.metadata.json` file). You can easily check if your module decorator is present by greping the name of it inside a metadata file.

As far as I know, decorator metadata was missing because of different module resolution for typescript compilation and metadata generation (see comment: https://github.com/dherges/ng-packagr/issues/382#issuecomment-362318700).

TL;DR: Try to change `export` statements in your library:
```
// from
export * from './module-directory';
// to
export * from './module-directory/index';
```
to fix the issue with metadata module resolution

**Comment by srinivasganaparthi** [comment]
I have tried this solution but still no luck

**Comment by kkapoor-git** [comment]
any solution / recommendations how to fix this issue

**Comment by Gillardo** [comment]
any update on this??

**Comment by divyanshidm** [comment]
I am also getting the same error. Any fixes?
